src/common/hobject.cc

```diff
@@ -80,33 +84,27 @@ set<string> hobject_t::get_prefixes(
 
 string hobject_t::to_str() const
 {
-  string out;
-
-  char snap_with_hash[1000];
-  char *t = snap_with_hash;
-  const char *end = t + sizeof(snap_with_hash);
-
   uint64_t poolid(pool);
-  t += snprintf(t, end - t, "%.*llX", 16, (long long unsigned)poolid);
-
   uint32_t revhash(get_nibblewise_key_u32());
-  t += snprintf(t, end - t, ".%.*X", 8, revhash);
-
-  if (snap == CEPH_NOSNAP)
-    t += snprintf(t, end - t, ".head");
-  else if (snap == CEPH_SNAPDIR)
-    t += snprintf(t, end - t, ".snapdir");
-  else
-    t += snprintf(t, end - t, ".%llx", (long long unsigned)snap);
 
-  out.append(snap_with_hash, t);
+  string out = [=, this]() {
+    if (snap == CEPH_NOSNAP) {
+      return fmt::format(FMT_COMPILE("{:016X}.{:08X}.head."), poolid, revhash);
+    } else if (snap == CEPH_SNAPDIR) {
+      return fmt::format(
+	  FMT_COMPILE("{:016X}.{:08X}.snapdir."), poolid, revhash);
+    } else {
+      return fmt::format(
+	  FMT_COMPILE("{:016X}.{:08X}.{:X}."), poolid, revhash,
+	  (unsigned long long)snap);
+    }
+  }();
```

Comment:
nit: This works, although I think that we should avoid the complexity here and use a switch case instead.

Reply:
Agree. Modified to assigning into an empty out (I don't think a switch case adds clarity here, and the code produced should be the same).